harvester_e2e_tests/integration/test_vm_snapshot.py

```diff
@@ -0,0 +1,556 @@
+
+    @pytest.mark.dependency(depends=["source_vm_snapshot"])
+    def test_restore_into_new_vm_from_vm_snapshot(self, api_client,
+                                                  restored_from_snapshot_vm, source_vm,
+                                                  ssh_keypair, host_shell,
+                                                  vm_shell, wait_timeout):
+        """
+        Test that restoring the `vm-snapshot` into a new virtual
+        machine results in a virtual machine with the expected
+        well-known file (`test.txt`) with the expected contents
+        (`123`).
+
+        Prerequisites:
+        1. The source VM from the first test case and its
+           snapshot (`vm-snapshot`).
+        2. A VM restored from `vm-snapshot`.
+        """
+
+        _, ssh_user = source_vm
+
+        def actassert(sh):
+            out, _ = sh.exec_command("cat test.txt")
+            assert "123" in out
+
+        vm_shell_do(restored_from_snapshot_vm, api_client,
+                    host_shell, vm_shell,
+                    ssh_user, ssh_keypair,
+                    actassert, wait_timeout)
+
+    @pytest.mark.dependency(depends=["source_vm_snapshot"])
+    def test_replace_is_rejected_when_deletepolicy_is_retain(self, api_client,
+                                                             source_vm, vm_snapshot_name,
+                                                             wait_timeout):
+        name, _ = source_vm
+
+        stop_vm(name, api_client, wait_timeout)
+
+        """
+        Test that the Harvester API rejects a `replace`
+        VirtualMachineRestore where the deletePolicy is
+        not `retain`.
+
+        Prequisites:
+        1. The original VM (`source-vm`) and snapshot (`vm-snapshot`)
+        from the first test case.
+        """
+        spec = api_client.backups.RestoreSpec.for_existing(delete_volumes=True)
+        code, data = api_client.vm_snapshots.restore(vm_snapshot_name, spec)
+
+        reason = data.get("message")
+
+        wantmsg = "Delete policy with backup type snapshot"
+        " for replacing VM is not supported"
+
+        assert wantmsg in reason
+        assert 422 == code
+
+    @pytest.mark.dependency(name="replaced_source_vm", depends=["source_vm_snapshot"])
+    def test_replace_vm_with_vm_snapshot(self, api_client,
+                                         source_vm, vm_snapshot_name,
+                                         ssh_keypair, host_shell,
+                                         vm_shell, wait_timeout):
+        """
+        Test that the original virtual machine can be replaced
+        from its original snapshot (`vm-snapshot`) and that
+        the snapshot's data contains the well-known file (`test.txt`)
+        and its expected contents (`123`).
+
+        Prerequisites:
+        `vm-snapshot` VM snapshot exists.
+        """
+        name, ssh_user = source_vm
+
+        stop_vm(name, api_client, wait_timeout)
```

Comment:
I think the implementation is still not fit the test criteria, data in the `source_vm` is the same as the snapshot taken, so we can't say that the restore is success or not.

what we test a snapshot is like:
1. Create `vm1`
2. Write data into `vm1`, naming its state as `a`
3. Take snapshot `vm1-s` for `vm1`, so the snapshot is in same state `a` of the `vm1`
4. Write or delete data into `vm1`, naming its state as `b`
5. Restore snapshot `vm1-s` for `vm1`
6. `vm1` should back to state `a`

Steps 4 is the criteria for the test case, then we can guarantee that restore the snapshot is succeed because the state is changed.

For this implementation, Step 4 is missing, so we can't guarantee that the snapshot is restored or not.

Reply:
We will not make the side effect in the test case `test_vm_snapshot_create`, the test case is purely require that create a snapshot, modify it after snapshot is not included.

Without the side effect, when we restore the VM into **new** one, we can simply compare files in both VMs to make sure it restored correctly.

And If we are going to verify restore existing VM, we would update the VM before restore to make sure what we changed will be reverted.

Based on this, we will not care about what we wrote into, so there will not have some constant assertion like `assert "123" in out`, because we will compare files in two VMs or check the modification is reverted.

Reply:
> We will not make the side effect in the test case test_vm_snapshot_create,

The code currently _does_ make this side-effect. Just to make sure I'm parsing your comment correctly, are you saying that you want me to remove that side-effect from the test case and then duplicate it across all of the other relevant test cases? I put it there initially because I didn't want every test case to have a connect, write, wait 2-seconds loop, which would extend the runtime of the test suite.

Reply:
yes it is necessary.
When we are just execute `test_vm_snapshot_create`, what we will expects is, the VM having the same state as the snapshot been taken, modify the VM is not required and not expected.

For other test cases, we will not depends on the _side effect_ of data what we wrote, in **restore new**, we would compare two VMs having same files and content, not compare the constants from the `test_vm_snapshot_create`; in **restore existing**, we will just check what we wrote/deleted before restore, will be reverted, not compare the constants from the `test_vm_snapshot_create`.

In current implementation, we would need make the filename comes from `source_vm`, then the implementation would be more stable to extend when we are going to change the data what we are going to wrote, we would just need to store file names and passing them in test cases.

Reply:
Sorry for the delay! 🙂 

I think I've addressed this in the latest revision. The new test fixtures will overwrite contents prior to restore and assert that the restored or replaced VM have the original contents prior to the overwrite.